File: web/src/lib/learning-engine/fuzzy-matcher.ts
export enum MatchResult {
    correct,
    typo,
    wrong,
}

export class FuzzyMatcher {
    checkAnswer(target: string, input: string): MatchResult {
        const cleanTarget = this._normalize(target);
        const cleanInput = this._normalize(input);

        if (cleanTarget === cleanInput) {
            return MatchResult.correct;
        }

        const distance = this._levenshtein(cleanTarget, cleanInput);

        // Tolerance: 
        // Length <= 3: 0 tolerance
        // Length <= 6: 1 tolerance
        // Length > 6: 2 tolerance
        let tolerance = 0;
        if (cleanTarget.length > 6) {
            tolerance = 2;
        } else if (cleanTarget.length > 3) {
            tolerance = 1;
        }

        if (distance <= tolerance) {
            return MatchResult.typo;
        }

        return MatchResult.wrong;
    }

    private _normalize(text: string): string {
        return text.trim().toLowerCase().replace(/[^\w\s]/g, '');
    }

    private _levenshtein(a: string, b: string): number {
        if (a.length === 0) return b.length;
        if (b.length === 0) return a.length;

        const matrix: number[][] = [];

        // Increment along the first column of each row
        for (let i = 0; i <= b.length; i++) {
            matrix[i] = [i];
        }

        // Increment each column in the first row
        for (let j = 0; j <= a.length; j++) {
            matrix[0][j] = j;
        }

        // Fill in the rest of the matrix
        for (let i = 1; i <= b.length; i++) {
            for (let j = 1; j <= a.length; j++) {
                if (b.charAt(i - 1) === a.charAt(j - 1)) {
                    matrix[i][j] = matrix[i - 1][j - 1];
                } else {
                    matrix[i][j] = Math.min(
                        matrix[i - 1][j - 1] + 1, // substitution
                        Math.min(
                            matrix[i][j - 1] + 1, // insertion
                            matrix[i - 1][j] + 1  // deletion
                        )
                    );
                }
            }
        }

        return matrix[b.length][a.length];
    }
}

File: mobile/lib/core/services/learning_engine/fuzzy_matcher.dart
import 'dart:math';

enum MatchResult { correct, typo, wrong }

class FuzzyMatcher {
  /// Checks the user input against the target word.
  /// Returns [MatchResult.correct] if exact match (case insensitive).
  /// Returns [MatchResult.typo] if Levenshtein distance is small.
  /// Returns [MatchResult.wrong] otherwise.
  static MatchResult check(String target, String input) {
    final t = target.trim().toLowerCase();
    final i = input.trim().toLowerCase();

    if (t == i) return MatchResult.correct;

    final distance = levenshtein(t, i);

    // Logic for "Typo":
    // If word is short (<= 3 chars), must be exact.
    // If word is medium (4-5 chars), allow 1 error.
    // If word is long (> 5 chars), allow 2 errors.

    if (t.length <= 3) return MatchResult.wrong;

    int allowedErrors = (t.length > 5) ? 2 : 1;

    if (distance <= allowedErrors) {
      return MatchResult.typo;
    }

    return MatchResult.wrong;
  }

  static int levenshtein(String a, String b) {
    // Standard Levenshtein implementation
    if (a == b) return 0;
    if (a.isEmpty) return b.length;
    if (b.isEmpty) return a.length;

    List<int> previousRow = List<int>.generate(b.length + 1, (i) => i);
    List<int> currentRow = List<int>.filled(b.length + 1, 0);

    for (int i = 0; i < a.length; i++) {
      currentRow[0] = i + 1;
      for (int j = 0; j < b.length; j++) {
        int cost = (a[i] == b[j]) ? 0 : 1;
        currentRow[j + 1] = min(
          currentRow[j] + 1, // Insertion
          min(
            previousRow[j + 1] + 1, // Deletion
            previousRow[j] + cost, // Substitution
          ),
        );
      }
      // Swap rows
      final temp = previousRow;
      previousRow = currentRow;
      currentRow = temp;
    }

    return previousRow[b.length];
  }
}
